src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -2767,17 +2767,18 @@ protected RubySymbol toSymBroken(Object string,
     @ImportStatic(StringGuards.class)
     public abstract static class ReverseBangNode extends CoreMethodArrayArgumentsNode {
 
-        @Child CharacterLengthNode characterLengthNode = CharacterLengthNode.create();
+        @Child
+        TruffleString.CodePointLengthNode codePointLengthNode = TruffleString.CodePointLengthNode.create();
         @Child private TruffleString.FromByteArrayNode fromByteArrayNode = TruffleString.FromByteArrayNode.create();
 
-        @Specialization(guards = "reverseIsEqualToSelf(string, characterLengthNode)")
+        @Specialization(guards = "reverseIsEqualToSelf(string.getTString(), codePointLengthNode, string.encoding)")
         protected RubyString reverseNoOp(RubyString string) {
             return string;
         }
 
         @Specialization(
                 guards = {
-                        "!reverseIsEqualToSelf(string, characterLengthNode)",
+                        "!reverseIsEqualToSelf(string.getTString(), codePointLengthNode, string.encoding)",
```

Comment:
I think you're better off having `reverseIsEqualToSelf` accept `RubyString` like it used to and have that method pull out the relevant fields. It'd simplify the guards in this node.